### ISSUE TITLE
Add ability for per-sample weights in regression problems

### ISSUE THREAD

**Issue by jamescolless** [dreamquark-ai/tabnet#219]
## Feature request

**What is the expected behavior?**

Add the ability to use per-sample weights in regression problems.

**What is motivation or use case for adding/changing the behavior?**

For many regression problems we care more about getting certain predictions right than others. This is a common feature of all other ML models (xgboost, lgbm, H20 etc. etc.)

**How should this be implemented in your opinion?**

When training the network we need to calculate the appropriately weighted loss to then backprop through.

**Are you willing to work on this yourself?**

yes if I knew where to start!

**Comment by brunovilar** [comment]
To be honest, I almost suggested the same feature 2 weeks ago, but for classification. I noticed that someone also suggested the [same recently](https://github.com/dreamquark-ai/tabnet/issues/216). I did not post it because, while I was checking the code to see what to change, I noticed that the code already considers the option to provide instance weights as a list. 

Maybe it is the case of updating the documentation, as, for now, [it only says](https://github.com/dreamquark-ai/tabnet/blob/master/pytorch_tabnet/abstract_model.py#L105):

>             weights : bool or dictionnary
>              0 for no balancing
>              1 for automated balancing
>              dict for custom weights per class


As far as I could understand, the way the instance weights are working now is based on the probability of sampling the weights when creating the [data loaders](https://github.com/dreamquark-ai/tabnet/blob/d216fbfa4dadd6c8d4110fa8da0f1c0bdfdccc2d/pytorch_tabnet/utils.py). I believe it simplifies the implementation a lot. However, does this solution is as good as really providing a weight to each instance and assuring that all instances are really used by the model when training? Using all instances, each of them using a specific weight, at the evaluation could also change a little the results in comparison to just sampling the instances. I just don't know if it is worth doing all modifications for this change. I am just conjecturing...

**Comment by Optimox** [comment]
@brunovilar we implemented a very standard way of dealing with class imbalances.

There is no fundamental difference between changing the sampling weights and changing the loss weight.
It's just two ways to deal with the same problem, we opted for the simplest one.

However, one can easily customize the loss function in order to give importance to specific weights.

- If the weights are directly related to the target (as @brunovilar  seem to be willing to do for classification), then you have access to the target at each batch and can weight your loss as you want. It seems fairly easy to do for experts that don't want to use sampling techniques but other techniques. By the way if you look at the documentation for TabNetRegressor you'll find the correct info about weights.

- If it's related to the sample themselves it's going to be a bit more difficult, and honesty I think that changing sampling weights is the safest way to do this.

@brunovilar @jamescolless could you please give concrete examples of what kind of problems you are trying to solve, and what prevents you from solving them with TabNet? TabNet has a certain way of dealing with class imbalance and it does not have to match how XGBoost is doing (by the way H2O is probably doing something different for class imbalance for every model they are using so I'm not sure it's relevant to compare any specific library to H2O). I would be happy to add a new feature, but I first need to understand the problem it solves.

**Comment by brunovilar** [comment]
@Optimox, for me, the current implementation is already enough. I mentioned the documentation for classification because it seems to only provide information for non-personalized instance weights. I could only identify that TabNet already provides what I need after checking the code and since this doubt is happening "frequently", maybe updating the documentation allows other people to identify this option faster and without creating a new feature request.

Another option is to make API more similar to the one used Sklearn. But for that, it would be necessary to create a new parameter (_sample_weights_), which could create conflicting configurations (like having, both _weight_ and _sample_weights_ defined at the same time).

**Comment by jamescolless** [comment]
Ah so perhaps this is indeed done and only the documentation confused me! To be clear, if I pass a numpy array of weights when I call model.fit() for a regression problem, it will then calculate the *weighted*  mse and optimize on that quantity?

**Comment by Optimox** [comment]
@brunovilar yes the documentation should probably be improved, I'll try to work on it, thanks!

@jamescolless yes and no. You can add per sample weights for regression, but it won't computed a weighted MSE, instead it will randomly sample every line according to the given weights. So if you have a very large training set it would be equivalent to weighted MSE starting from first epoch, otherwise in the end after a few epochs it would be similar to weighted MSE as your important sample will appear more frequently. So it's not strictly the same as weighted MSE, but the loss with weights will indeed try to perform better on important samples.
About the `eval_metric`, you have access to any eval_set (could be train, valid or test or anything) and the entire predictions of the network, so you can easily compute the exact weigthed MSE that you want to optimise (you'll just need to hard code the weights inside your evaluation metric - different weights for every eval set).

Let me know if it's still unclear!

**Comment by jamescolless** [comment]
Nope this all makes sense thanks!
